cloud-exec.md

```diff
@@ -0,0 +1,153 @@
+
+## Monitoring 
+
+Additional features might be needed to improve user experienc:
+1. **State monitoring.** A user should see if an instance was already
+   allocated, what stage is running.
+2. **Metrics monitoring.**  Latest metrics monitoring (with some reasonable
+   delay - 10-30-60 seconds)
+
+
+## Resource orchestrators
+
+1. AWS
+2. Azure
+3. GCP (optional)
+4. Remote over SSH
+5. Kubernetes (K8S)
+6. HashiCorp Nomad (optional)
+7. Container services (ECS and Azure/GCP analogs)
+
+
+# Detailed design
+
+## Iterative Terraform provider
+
+The cloud management part seems like a heavy part of the project (from
+  the implementation point of view). However, the problem of executing and
+  managing cloud instances was solved in the CML project.
+
+CML separates out the user API part (commands) and the cloud instace
+  management code part.
+The cloud management is extracted to a separate **Iterative Terraform (TF)
+  provider**:
+  https://github.com/iterative/terraform-provider-iterative
+
+The Iterative provider is specific to data scientists, not DevOps (TF's
+  target audience). It has a bit simplified API and some ML specific
+  functionality (some of these are coming) like: recover spot-instance if
+  it was deleted or attach a reusable drive.
+
+The Iteratve TF provider can be used from DVC for cloud instace management.
+Some Python libraries might be required  for access to GoLang specific
+terraform-provider.
+
+Today the provider supports:
+- AWS, Azure, K8S orchestrators
+- Spot instances and Volumes optimizations.
+A few of the initial scenarios can be already implemented with these
+functionalities. More functionality is comming.
+
+## Instances configuration
+
+Instance configuration is a definition of an instance and its behavior:
+- orchestrator: AWS, GCP, Azure, remote, K8S, ...
+- instance type
+- volume
+- many other options
+
+Instance configuration is well developed in CML project and can be reused.
+
+## Executors and naming
+
+Many instance types and even clouds could be used in a single DVC project.
+We need a way of configuring **executors** and reusing them the same way
+  as user can configure a data remote.
+It can look like this:
+```bash
+$ dvc exp run --executor my-gpu-tesla
+...
+$ dvc exp run --executor my-gpu-tesla8
+...
+$ dvc exp run --executor tesla8-ram
+```
+
+# How We Teach This
+
+We need to introduce a concept **Executor** to DVC.
+This convept should not break the compatibility and should not require
+significant changes in the docs (except command options).v
+However, it will require a new section in the docs to explain the cloud
+execution for the users who'd like to use this feature.
+
+# Drawbacks
+
+This feature opens new user scenarios which increase the complexity of DVC,
+it will require new sections in the docs, etc.
+It might be beneficial to extract as much of this functionality as possible to
+external products and tools (terraform providers for example). 
```

Comment:
The doc opens with "Users need to run ML experiments in cloud instances or remote machines. DVC should support this out of the box." but this part seems to question that. My first Q is to clarify or decide on that. Do we want `dvc` CLI to have remote execution?

If it was possible to elaborate a little on a hypothetical external tool as alternative (e.g. Terraform provider) that could help. And what about putting it into `dvcx` (once installed could extend the `dvc` CLI) or in another (perhaps GUI-based) tool from Iterative?

Reply:
Modified the intro and this section.

Reply:
Thanks. I'd resolve this (and others) but I can't on this repo.